src/common/libpmi/pmi2.c

```diff
@@ -248,10 +254,43 @@ int PMI2_Info_GetSize (int *size)
     return PMI2_FAIL;
 }
 
+/* Cray MPI: look up a node-scope key stored with PMI2_Info_PutNodeAttr().
+ * If waitfor is nonzero, try once per second until the key is available.
+ */
 int PMI2_Info_GetNodeAttr (const char name[],
                            char value[], int valuelen, int *found, int waitfor)
 {
-    return PMI2_FAIL;
+    const char *kvsname;
+    int result;
+    int tries = 0;
+    char local_name[PMI2_MAX_KEYLEN + 8];
+
+    if (!name || !value)
+        return PMI2_ERR_INVALID_ARG;
+    result = get_cached_kvsname (pmi_global_ctx, &kvsname);
+    if (result != PMI2_SUCCESS)
+        return result;
+    if (snprintf (local_name,
+                  sizeof (local_name),
+                  "local::%s", name) >= sizeof (local_name))
+        return PMI2_ERR_INVALID_KEY_LENGTH;
+    do {
+        if (tries++ > 0)
+            sleep (1);
+        result = pmi_simple_client_kvs_get (pmi_global_ctx,
+                                            kvsname,
+                                            local_name,
+                                            value,
+                                            valuelen);
+        if (result != PMI2_ERR_INVALID_KEY && result != PMI2_SUCCESS)
+            return result;
+        tries++;
+    } while (result == PMI2_ERR_INVALID_KEY && waitfor != 0);
```

Comment:
Probably best to have a limit to this to prevent endlessly looking for an invalid key?

Reply:
The reason I did it that way was this comment in argonne's pmi2.h:
> If waitfor is non-zero, the function will never return with found set to zero.

Meh, I agree with you though.  I'll add a 30s timeout.